File: plat/js/entities/bird.ts
import * as SLT from 'slate2d';
import * as Draw from 'draw';
import * as Assets from 'assets';

import Entity from '../entity.js';
import Dir from '../dir.js';
import CollisionType from '../collisiontype.js';
import { World } from '../game.js';
import { Player } from './player.js';
import { dbg, dbgval } from '../printwin.js';

class Bird extends Entity {
  type = 'Bird';
  drawOfs: [number, number] = [-1, -2];
  sprite = Assets.find('bird');
  delta: [number, number] = [0, 0];
  dir: [number, number] = [0, 0];
  moveAmt: [number, number] = [0, 0];

  constructor(args: { [key: string]: any }) {
    super(args);

    // points are kinda hard to work with and they're not in any format
    // that maps cleanly to the entity position
    const dest = args.properties?.Destination ?? this.pos;
    const tilePos = [Math.floor(this.pos[0] / 16), Math.floor(this.pos[1] / 16)];

    this.delta[0] = (dest[0] - tilePos[0]) * 16;
    this.delta[1] = (dest[1] - tilePos[1]) * 16;

    const absDelta = [Math.abs(this.delta[0]), Math.abs(this.delta[1])];

    let dx = absDelta[0] > absDelta[1] ? 1 : absDelta[0] / absDelta[1];
    let dy = absDelta[1] > absDelta[0] ? 1 : absDelta[1] / absDelta[0];

    dx *= Math.sign(this.delta[0]) * 0.5;
    dy *= Math.sign(this.delta[1]) * 0.5;

    this.dir = [dx, dy];
    this.worldCollide = false;
  }

  die() {
    super.die();
    World().spawnDeathParticle(this, 3);
  }

  canCollide(other: Entity, dir: Dir) {
    if (other instanceof Player && dir == Dir.Up) return CollisionType.Platform;
    else return CollisionType.Trigger;
  }

  update(ticks: number, dt: number) {
    this.frame = Math.floor(ticks / 8 % 4);
    this.frame = this.frame == 3 ? 1 : this.frame;

    this.moveX(this.dir[0]);
    this.moveY(this.dir[1]);
    this.moveAmt[0] += this.dir[0];
    this.moveAmt[1] += this.dir[1]

    const dim = this.delta[0] > this.delta[1] ? 0 : 1;
    if (Math.abs(this.moveAmt[dim]) >= Math.abs(this.delta[dim])) {
      this.dir[0] *= -1;
      this.dir[1] *= -1;
      this.moveAmt = [0, 0];
    }
  }

  collide(other: Entity, dir: Dir) {
    this.handlePlayerStomp(other, dir);
  }
}

export { Bird };

File: csgame/entities/Bird.cs
using Slate2D;

[Spawnable]
class Bird : Entity
{
    (int X, int Y) Delta = (0, 0);
    (float X, float Y) MoveDir = (0, 0);
    (float X, float Y) MoveAmt = (0, 0);

    public Bird(LDTKEntity ent) : base(ent)
    {
        Sprite = Assets.Find("bird");
        DrawOfs = (-1, -2);

        (int X, int Y) dest = ((int X, int Y))((ent.Properties["Destination"]?.Point) ?? (Pos.X, Pos.Y));
        (int X, int Y) tilePos = ((int X, int Y))(MathF.Floor(Pos.X / 16), MathF.Floor(Pos.Y / 16));

        Delta.X = (dest.X - tilePos.X) * 16;
        Delta.Y = (dest.Y - tilePos.Y) * 16;

        (float X, float Y) absDelta = (Math.Abs(Delta.X), Math.Abs(Delta.Y));

        float dx = absDelta.X > absDelta.Y ? 1 : absDelta.X / absDelta.Y;
        float dy = absDelta.Y > absDelta.X ? 1 : absDelta.Y / absDelta.X;

        dx *= Math.Sign(Delta.X) / 2f;
        dy *= Math.Sign(Delta.Y) / 2f;

        MoveDir = (dx, dy);
        WorldCollide = false;
    }

    public override void Die()
    {
        base.Die();
        Main.World.SpawnDeathParticle(this, 3);
    }

    public override CollisionType CanCollide(Entity other, Dir dir)
    {
        if (other is Player && dir == Dir.Up) return CollisionType.Platform;
        else return CollisionType.Trigger;
    }

    public override void Update(uint ticks, float dt)
    {
        Frame = ticks / 8 % 4;
        Frame = Frame == 3 ? 1 : Frame;

        MoveX(MoveDir.X);
        MoveY(MoveDir.Y);
        MoveAmt.X += MoveDir.X;
        MoveAmt.Y += MoveDir.Y;

        var dim = Delta.X > Delta.Y ? 0 : 1;
        ref var moveAmt = ref (dim == 0 ? ref MoveAmt.X : ref MoveAmt.Y);
        ref var delta = ref (dim == 0 ? ref Delta.X : ref Delta.Y);

        if (Math.Abs(moveAmt) >= Math.Abs(delta))
        {
            MoveDir.X *= -1;
            MoveDir.Y *= -1;
            MoveAmt = (0, 0);
        }
    }

    public override void Collide(Entity other, Dir dir)
    {
        HandlePlayerStomp(other, dir);
    }

}